# Fase 1 — Exploración y Diagnóstico (EDA)
## Dataset: ABC Corporation — HR Analytics

**Objetivo:** Entender el dataset antes de tocar nada.
- **Pareja A:** estructura, calidad, nulos, duplicados y estadísticas numéricas.
- **Pareja B:** contenido, variable objetivo y columnas categóricas.

---

## Librerías e importaciones

In [1]:
# Librerías necesarias para el análisis exploratorio
import pandas as pd

# Mostrar todas las columnas al imprimir DataFrames
pd.set_option('display.max_columns', None)

## Carga del dataset y comprobación de integridad

In [2]:
# Cargamos el dataset original
df = pd.read_csv('hr_clean.csv')
 

# 1. Validamos las dimensiones exactas del dataset
print(f"¿El número de filas es el esperado? {'Sí' if df.shape[0] == 1474 else 'No'} ({df.shape[0]} filas)")
print(f"¿El número de columnas es el esperado? {'Sí' if df.shape[1] == 35 else 'No'} ({df.shape[1]} columnas)")

# 2. Verificamos que no haya problemas de lectura
print(f"\nTotal de celdas con datos: {df.notnull().sum().sum()}")
print(f"Total de celdas vacías (nulos): {df.isnull().sum().sum()}")

# 3. Inspección visual rápida de las primeras 3 filas
print("\n--- Vista de control de las 3 primeras filas ---")
display(df.head(3))

¿El número de filas es el esperado? No (1470 filas)
¿El número de columnas es el esperado? No (32 columnas)

Total de celdas con datos: 47040
Total de celdas vacías (nulos): 0

--- Vista de control de las 3 primeras filas ---


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,1,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,1,11,3,1,0,8,0,1,6,4,0,5
1,49,0,Travel_Frequently,279,Research & Development,8,1,Life Sciences,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,0,23,4,4,1,10,3,3,10,7,1,7
2,37,1,Travel_Rarely,1373,Research & Development,2,2,Other,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,1,15,3,2,0,7,3,3,0,0,0,0


---
## Tarea A1 — ¿Qué forma tiene el dataset?

Mirar cuántas filas/columnas hay y el tipo de dato de cada una.

In [3]:
# 1. Cuántas filas y columnas hay
print('Filas    :', df.shape[0])
print('Columnas :', df.shape[1])

# 2. Conteo de tipos de datos
print('\nCONTEO DE TIPOS DE DATOS:')
print(df.dtypes.value_counts())

# 3. Qué tipo de dato tiene cada columna en detalle
print('\nTIPOS DE DATOS POR COLUMNA:')
print(df.dtypes.to_string())

# 4. Ver las primeras 10 filas
df.head(10)

Filas    : 1470
Columnas : 32

CONTEO DE TIPOS DE DATOS:
int64    26
str       6
Name: count, dtype: int64

TIPOS DE DATOS POR COLUMNA:
Age                         int64
Attrition                   int64
BusinessTravel                str
DailyRate                   int64
Department                    str
DistanceFromHome            int64
Education                   int64
EducationField                str
EmployeeNumber              int64
EnvironmentSatisfaction     int64
Gender                        str
HourlyRate                  int64
JobInvolvement              int64
JobLevel                    int64
JobRole                       str
JobSatisfaction             int64
MaritalStatus                 str
MonthlyIncome               int64
MonthlyRate                 int64
NumCompaniesWorked          int64
OverTime                    int64
PercentSalaryHike           int64
PerformanceRating           int64
RelationshipSatisfaction    int64
StockOptionLevel            int64
TotalWorkingYe

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,1,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,1,11,3,1,0,8,0,1,6,4,0,5
1,49,0,Travel_Frequently,279,Research & Development,8,1,Life Sciences,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,0,23,4,4,1,10,3,3,10,7,1,7
2,37,1,Travel_Rarely,1373,Research & Development,2,2,Other,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,1,15,3,2,0,7,3,3,0,0,0,0
3,33,0,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,1,11,3,3,0,8,3,3,8,7,3,0
4,27,0,Travel_Rarely,591,Research & Development,2,1,Medical,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,0,12,3,4,1,6,3,3,2,2,2,2
5,32,0,Travel_Frequently,1005,Research & Development,2,2,Life Sciences,8,4,Male,79,3,1,Laboratory Technician,4,Single,3068,11864,0,0,13,3,3,0,8,2,2,7,7,3,6
6,59,0,Travel_Rarely,1324,Research & Development,3,3,Medical,10,3,Female,81,4,1,Laboratory Technician,1,Married,2670,9964,4,0,20,4,1,3,12,3,2,1,0,0,3
7,30,0,Travel_Rarely,1358,Research & Development,24,1,Life Sciences,11,4,Male,67,3,1,Laboratory Technician,3,Divorced,2693,13335,1,0,22,4,2,1,1,2,3,1,0,0,0
8,38,0,Travel_Frequently,216,Research & Development,23,3,Life Sciences,12,4,Male,44,2,3,Manufacturing Director,3,Single,9526,8787,0,0,21,4,2,0,10,2,3,9,7,1,8
9,36,0,Travel_Rarely,1299,Research & Development,27,3,Life Sciences,13,3,Male,94,3,2,Healthcare Representative,3,Married,5237,16577,6,0,13,3,2,2,17,3,2,7,7,7,3


**Hallazgos Tarea A1:**
- El dataset tiene **1474 empleados** y **35 variables**
- 20 columnas `int64`, 9 `object` (texto), 6 `float64`
- Columnas mal tipadas — son `float64` pero deberían ser `int`:
  - `Age` → 41.0 debería ser 41
  - `MonthlyIncome` → debería ser int
  - `JobSatisfaction` → escala 1-4, no necesita decimales
  - `YearsWithCurrManager` → los años son enteros
  - `TrainingTimesLastYear` → los entrenamientos son enteros
  - `StandardHours` → siempre 80.0 → además inútil 🗑️
- Ya se ven NaN a simple vista en las primeras 10 filas
- Se arreglará todo en la **Fase 2**

---
## Tarea A2 — ¿Hay huecos vacíos?

Buscar columnas con valores NaN y contar cuántos hay.

In [4]:
# 1. Contar nulos por columna
print('NULOS POR COLUMNA:')
nulos = df.isnull().sum()
# df.isnull() → tabla True/False; .sum() → suma los True de cada columna
# nulos[nulos > 0] → solo columnas con algún nulo
print(nulos[nulos > 0].sort_values(ascending=False))

# 2. Porcentaje de nulos por columna
print('\nPORCENTAJE DE NULOS:')
# .mean() calcula la proporción de True/False → * 100 lo convierte en porcentaje
porcentaje = df.isnull().mean() * 100
porcentaje = porcentaje[porcentaje > 0].round(2).sort_values(ascending=False)
print(porcentaje)

NULOS POR COLUMNA:
Series([], dtype: int64)

PORCENTAJE DE NULOS:
Series([], dtype: float64)


**Hallazgos Tarea A2:**
- **11 columnas** tienen valores nulos de 35 totales

🔴 **Más del 10% — graves:**
- `StandardHours` → 164 nulos (11.13%) → además es constante → eliminar en Fase 2 🗑️
- `YearsWithCurrManager` → 148 nulos (10.04%) → rellenar con mediana

🟡 **Entre 5% y 10% — moderados:**
- `MaritalStatus` → 132 nulos (8.96%) → además tiene la errata `Marreid` → rellenar con moda
- `BusinessTravel` → 117 nulos (7.94%) → rellenar con moda
- `TrainingTimesLastYear` → 88 nulos (5.97%) → rellenar con mediana
- `Age` → 73 nulos (4.95%) → rellenar con mediana

🟢 **Menos del 5% — leves:**
- `EducationField` → 58 nulos (3.93%) → rellenar con moda
- `OverTime` → 44 nulos (2.99%) → rellenar con moda
- `Department` → 29 nulos (1.97%) → rellenar con moda
- `JobSatisfaction` → 29 nulos (1.97%) → rellenar con mediana
- `MonthlyIncome` → 14 nulos (0.95%) → rellenar con mediana

---
## Tarea A3 — ¿Hay filas repetidas?

Detectar si hay empleados duplicados en el dataset.

In [5]:
# 1. Contar filas duplicadas
# df.duplicated() marca como True las filas exactamente iguales a otra anterior
duplicados = df.duplicated().sum()
print('Filas duplicadas:', duplicados)

# 2. Ver cómo son esas filas duplicadas
# keep=False marca TODAS las copias (original + duplicado) para verlas juntas
if duplicados > 0:
    print('\nAsí son las filas duplicadas:')
    display(df[df.duplicated(keep=False)])

Filas duplicadas: 0


**Hallazgos Tarea A3:**
- Hay **4 filas duplicadas** en el dataset
- Las copias están al final del archivo (filas 1470-1473)
  → probablemente error al añadir datos
- Se eliminarán en la **Fase 2** con `df.drop_duplicates()`
- Tras eliminarlas quedarán **1470 empleados** en lugar de 1474

---
## Tarea A4 — Estadísticas numéricas

Ver mínimo, máximo, media y desviación de columnas numéricas.

In [6]:
# 1. Estadísticas generales de todas las columnas numéricas
# .T transpone la tabla para que sea más legible
print('ESTADÍSTICAS COLUMNAS NUMÉRICAS:')
display(df.describe().T)

# 2. Columnas más relevantes para el proyecto
print('\nCOLUMNAS CLAVE — detalle:')
columnas_clave = ['Age', 'MonthlyIncome', 'YearsAtCompany',
                  'TotalWorkingYears', 'JobSatisfaction', 'WorkLifeBalance']

# Iteramos por cada columna clave para mostrar sus estadísticas principales
for columna in columnas_clave:
    minimo  = df[columna].min()
    maximo  = df[columna].max()
    media   = df[columna].mean().round(2)
    mediana = df[columna].median()
    print(f'\n{columna}:')
    print(f'  Min: {minimo}  |  Max: {maximo}  |  Media: {media}  |  Mediana: {mediana}')

ESTADÍSTICAS COLUMNAS NUMÉRICAS:


,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.895918,8.869106,18.0,31.00,36.0,42.00,60.0
Attrition,1470.0,0.161224,0.367863,0.0,0.00,0.0,0.00,1.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0



COLUMNAS CLAVE — detalle:

Age:
  Min: 18  |  Max: 60  |  Media: 36.9  |  Mediana: 36.0

MonthlyIncome:
  Min: 1009  |  Max: 19999  |  Media: 6483.52  |  Mediana: 4907.0

YearsAtCompany:
  Min: 0  |  Max: 40  |  Media: 7.01  |  Mediana: 5.0

TotalWorkingYears:
  Min: 0  |  Max: 40  |  Media: 11.28  |  Mediana: 10.0

JobSatisfaction:
  Min: 1  |  Max: 4  |  Media: 2.74  |  Mediana: 3.0

WorkLifeBalance:
  Min: 1  |  Max: 4  |  Media: 2.76  |  Mediana: 3.0


**Hallazgos Tarea A4:**
- **Salario:** media $6,497 pero mediana $4,907
  → la media está inflada por salarios muy altos (outliers)
  → la mediana es más representativa
- **Edad:** entre 18 y 60, media 36.9 → plantilla joven
- **Años en empresa:** mediana 5 años → la mitad lleva menos de 5 años → plantilla relativamente nueva
- **Satisfacción laboral:** media 2.73/4 → tirando a baja ⚠️
- **WorkLifeBalance:** media 2.76/4 → mejorable ⚠️
- `EmployeeCount` y `StandardHours`: desviación estándar = 0 → inútiles → eliminar en Fase 2

---
## Tarea B1 — ¿Cuántos empleados se fueron? (Attrition)

Analizar la columna `Attrition` (Yes/No), la variable más importante.

In [7]:
# Contamos cuántos empleados se fueron (Yes) y cuántos se quedaron (No)
df['Attrition'].value_counts()


Attrition
0    1233
1     237
Name: count, dtype: int64

In [8]:
# Calculamos el porcentaje de cada valor
# normalize=True da proporciones (0-1); * 100 las convierte en porcentaje
df['Attrition'].value_counts(normalize=True).mul(100).round(2)

Attrition
0    83.88
1    16.12
Name: proportion, dtype: float64

**Conclusión:** El 16.15% de los empleados han abandonado la empresa. Aunque no es una crisis absoluta, confirma una fuga de talento real. Además hay un **desbalanceo de clases** (83.85% No vs 16.15% Yes), algo importante a tener en cuenta para un modelo predictivo.

---
## Tarea B2 — ¿Qué valores únicos hay en las columnas categóricas?

Revisar columnas categóricas y detectar textos raros o errores.

In [9]:
# Revisamos los valores únicos de las columnas categóricas clave
# para detectar errores de formato, erratas o valores inesperados
columnas_categoricas = ['JobRole', 'Department', 'Gender', 'MaritalStatus', 'OverTime']

for col in columnas_categoricas:
    print(f'--- Valores únicos en la columna: {col} ---')
    # value_counts(dropna=False) muestra también los nulos
    print(df[col].value_counts(dropna=False))
    print()

--- Valores únicos en la columna: JobRole ---
JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: count, dtype: int64

--- Valores únicos en la columna: Department ---
Department
Research & Development    968
Sales                     439
Human Resources            63
Name: count, dtype: int64

--- Valores únicos en la columna: Gender ---
Gender
Male      882
Female    588
Name: count, dtype: int64

--- Valores únicos en la columna: MaritalStatus ---
MaritalStatus
Married     738
Single      436
Divorced    296
Name: count, dtype: int64

--- Valores únicos en la columna: OverTime ---
OverTime
0    1065
1     405
Name: count, dtype: int64



**Hallazgos detectados:**

- **`JobRole`:** Formato de texto incorrecto (ej. `' sALES eXECUTIVE '`) y espacios en blanco al inicio y al final.
- **`Department`:** Contiene 29 valores nulos. Hay empleados cuyo departamento se desconoce.
- **`MaritalStatus`:** Contiene 132 valores nulos y una errata (`'Marreid'` en vez de `'Married'`), lo que genera dos grupos para un mismo valor.
- **`OverTime`:** Contiene 44 valores nulos.
- **`Gender`:** Sin problemas detectados.

---
## Tarea B3 — ¿Hay columnas constantes?

Detectar columnas sin variabilidad (`EmployeeCount`, `Over18`, etc.).

In [10]:
# Contamos los valores únicos de cada columna ordenados de menor a mayor
# Las columnas con valor 1 son constantes y no aportan información al análisis
df.nunique().sort_values()

Attrition                      2
Gender                         2
OverTime                       2
PerformanceRating              2
Department                     3
MaritalStatus                  3
BusinessTravel                 3
JobInvolvement                 4
RelationshipSatisfaction       4
JobSatisfaction                4
StockOptionLevel               4
WorkLifeBalance                4
EnvironmentSatisfaction        4
Education                      5
JobLevel                       5
EducationField                 6
TrainingTimesLastYear          7
JobRole                        9
NumCompaniesWorked            10
PercentSalaryHike             15
YearsSinceLastPromotion       16
YearsWithCurrManager          18
YearsInCurrentRole            19
DistanceFromHome              29
YearsAtCompany                37
TotalWorkingYears             40
Age                           43
HourlyRate                    71
DailyRate                    886
MonthlyIncome               1336
MonthlyRat

**Hallazgos detectados:**

- **`EmployeeCount`**, **`Over18`** y **`StandardHours`** tienen un único valor único. No aportan información y se eliminarán en la Fase 2.
- **`EmployeeNumber`** es un identificador único por empleado (1470 valores distintos). No aporta valor analítico y también se puede eliminar.
- Se han detectado **4 filas duplicadas** en el dataset. Se eliminarán en la Fase 2.

---
## 📋 Reporte de hallazgos — Fase 1 (Equipo completo)

**1. Fuga de talento (variable objetivo):**
- El **16.15%** de los empleados han abandonado la empresa (`Attrition = Yes`).
- Hay un desbalanceo de clases (83.85% No vs 16.15% Yes) importante para un modelo predictivo.

**2. Problemas en columnas categóricas:**
- **`JobRole`:** Formato incorrecto (ej. `' sALES eXECUTIVE '`) y espacios en blanco al inicio y final.
- **`Department`:** Contiene 29 valores nulos.
- **`MaritalStatus`:** Contiene 132 valores nulos y una errata (`'Marreid'` en vez de `'Married'`).
- **`OverTime`:** Contiene 44 valores nulos.

**3. Problemas en columnas numéricas:**
- **Nulos detectados en:** `Age` (73), `JobSatisfaction` (29), `MonthlyIncome` (14), `TrainingTimesLastYear` (88), `YearsWithCurrManager` (148).
- **Tipos incorrectos:** `Age`, `JobSatisfaction`, `MonthlyIncome`, `TrainingTimesLastYear` y `YearsWithCurrManager` son `float64` pero deberían ser `int64`.

**4. Columnas constantes a eliminar:**
- `EmployeeCount`, `Over18` y `StandardHours` tienen un único valor. No aportan información.
- `EmployeeNumber` es un identificador único sin valor analítico. Se puede eliminar también.

**5. Filas duplicadas:**
- Se han detectado **4 filas duplicadas** (filas 1470-1473) que se eliminarán en la Fase 2.

**6. Estadísticas relevantes:**
- **Salario:** mediana $4,907 (más representativa que la media $6,497 inflada por outliers).
- **Satisfacción laboral:** media 2.73/4 → mejorable.
- **WorkLifeBalance:** media 2.76/4 → mejorable.

| Problema | Acción en Fase 2 |
|---|---|
| Columnas constantes | Eliminar `EmployeeCount`, `Over18`, `StandardHours` |
| Identificador único | Eliminar `EmployeeNumber` |
| Tipos float64 → int64 | Convertir tras imputar nulos |
| Nulos en numéricas | Imputar con **mediana** |
| Nulos en categóricas | Imputar con **moda** |
| `JobRole` con mayúsculas y espacios | `str.strip().str.title()` |
| `MaritalStatus` con errata 'Marreid' | Reemplazar por 'Married' |
| 4 filas duplicadas | `drop_duplicates()` |